### PR TITLE
Cudf-free shuffler tests

### PR DESCRIPTION
**Pull request by nirandaperera**
## Summary

The shuffler round-trip tests depended on cudf for input generation and result checking (`random_table_with_index`, `partition_and_pack`/`unpack_and_concat`, `CUDF_TEST_EXPECT_TABLES_EQUIVALENT`, `cudf::test::BaseFixture`), which made them heavy and coupled to the cudf integration. This replaces that with a self-contained, conservation-preserving data model built directly from `PackedData`, so the shuffler tests no longer depend on cudf and run as part of the default (non-`BUILD_CUDF_TESTS`) build.

### C++ utilities
- Add `ceil_div` and a `chunk_indices` view to `utils/misc.hpp`. `chunk_indices` tiles `[0, count)` into exactly `num_chunks` contiguous (possibly empty) ranges whose sizes sum to `count`.
- `ceil_div` is computed as `x / y + (x % y != 0)` rather than the naive `(x + y - 1) / y`, which overflows for large unsigned values and is UB on signed overflow near the type maximum.
- Add compile-time edge-case coverage for `ceil_div` in `test_misc.cpp` (`static_assert`s for exact/remainder division, zero numerator, denominator of one, and values at `uint64`/`int64` maxima).

### C++ shuffler tests (now cudf-free)
- Rework the shuffler tests (`test_shuffler.cpp`, `streaming/test_shuffler.cpp`) around `make_partition_data` (produces routed chunks) and `validate_partition_data` (verifies received chunks), dropping all `cudf`/`cudf_test` usage and switching to `rmm::` stream/MR helpers and `::testing::Test`.
- Generalize the test helpers `generate_packed_data`/`validate_packed_data` in `tests/utils.hpp` over the element type and build device buffers via the buffer resource + `cuda_memcpy_async` instead of cudf.
- De-couple `streaming/test_allgather.cpp` from `cudf_streaming/integrations/partition.hpp`.
- `cpp/tests/CMakeLists.txt`: move `test_shuffler.cpp`, `streaming/test_shuffler.cpp`, and `streaming/test_allgather.cpp` out of the `BUILD_CUDF_TESTS` block into the default build; the cudf block now only carries `test_partition.cpp`, `test_shuffler_many_streams.cpp`, and `streaming/test_leaf_actor.cpp`.

### Python (now cudf-free)
- Add a `PackedData.from_device_buffer(gpu_data, metadata, stream, br)` factory (plus `.pyi` stub) so packed data can be built from an rmm `DeviceBuffer` without cudf.
- Add shared, cudf-free shuffle test helpers to `rapidsmpf/testing.py`: `chunk_indices`, `generate_packed_data`, `validate_packed_data`, `make_partition_data`, and `validate_partition_data`.
- Rewrite the Python shuffler tests (`tests/test_shuffler.py`, `tests/streaming/test_shuffler.py`) around those helpers and the `Shuffler`/`ShufflerAsync` object interfaces, removing the `cudf` / `cudf_streaming` / `pylibcudf` dependencies.

Depends on https://github.com/rapidsai/rapidsmpf/pull/1087

### CODE REVIEW

File: cpp/include/rapidsmpf/memory/spill.hpp
Comment (pentschev) [inline cpp/include/rapidsmpf/memory/spill.hpp:1]: AFAICT, this is not public API anymore, although you still use it for cudf Python integration which seems appropriate to be removed. Move this to somewhere under `cpp/tests`.

File: cpp/tests/utils.hpp [hunk 1]
```diff
@@ -184,58 +185,70 @@ template <std::integral T = std::int64_t>
...
+    auto data = br.make_buffer(stream, std::move(reservation));
+
+    data->write_access([d_ptr = metadata->data(), m_size = metadata->size()](
+                           std::byte* ptr, rmm::cuda_stream_view op_stream
+                       ) {
+        RAPIDSMPF_CUDA_TRY(rapidsmpf::cuda_memcpy_async(ptr, d_ptr, m_size, op_stream));
+    });
 
-    return {std::move(metadata), br.move(std::move(data), stream)};
+    return {std::move(metadata), std::move(data)};
```
Comment (pentschev) [inline cpp/tests/utils.hpp:229]: Why is `stream` removed here? `write_access` does not sync the stream, so I think it's still needed for the consumer.

Comment (nirandaperera) [reply cpp/tests/utils.hpp:229]: @pentschev before, we were creating a `rmm::device_buffer` and "moving" it into buffer resource. IMO the better way here should have been make a reservation and then make buffer. This is done in L213 and it sets `stream` for the buffer. `op_stream == stream` in this case.

File: cpp/include/rapidsmpf/memory/spill.hpp
Comment (nirandaperera) [reply cpp/include/rapidsmpf/memory/spill.hpp:1]: I think Spill/UnSpillPartitions should reside in rapidsmpf. It has no dependency to cudf, so I moved it to a separate header.

File: cpp/include/rapidsmpf/utils/misc.hpp [hunk 2]
```diff
@@ -193,6 +195,49 @@ constexpr T safe_div(T x, T y) {
...
+ * non-negative and the denominator must be non-zero.
+ *
+ * @tparam T An integral type.
+ * @param x The numerator (must be non-negative).
+ * @param y The denominator (must be positive).
+ * @return T The ceiling of `x / y`.
+ */
+template <std::integral T>
+constexpr T ceil_div(T x, T y) {
+    return (x + y - 1) / y;
```
Comment (pentschev) [inline cpp/include/rapidsmpf/utils/misc.hpp]:
```suggestion
    return x / y + (x % y != 0);
```

Comment (pentschev) [reply cpp/include/rapidsmpf/utils/misc.hpp]: This is now public API, but the original overflows for large unsigned values and is UB for signed overflow, let's thus improve that. Add a test for edge cases too.

File: cpp/tests/test_shuffler.cpp [hunk 3]
```diff
@@ -285,112 +302,83 @@ INSTANTIATE_TEST_SUITE_P(
...
             shuffler,
             total_num_partitions,
-            [&](auto&& packed_chunks) { insert_fn(shuffler, std::move(packed_chunks)); },
-            [&]() { insert_finished_fn(shuffler); },
             100'000,  // total_num_rows
-            t_id,  // seed
-            cudf::hash_id::HASH_MURMUR3,
             stream,
             br.get()
         ));
```
Comment (pentschev) [inline cpp/tests/test_shuffler.cpp:400]: The original cuDF test used `t_id` as the seed, the new one does not pass a distinct base into `test_shuffler`, so cross-talk between concurrent shufflers can pass unnoticed. Please make data different for each thread.

File: cpp/tests/streaming/test_shuffler.cpp [hunk 4]
```diff
@@ -175,12 +78,111 @@ INSTANTIATE_TEST_SUITE_P(
...
+    // output chunk.
+    EXPECT_EQ(local_pids.size(), output_chunks.size());
+    const size_t n_ranks = static_cast<size_t>(comm->nranks());
+    for (auto& chunk : output_chunks) {
+        auto pid = chunk.sequence_number();
+        std::erase_if(local_pids, [pid](auto& p) { return p == pid; });
+
+        auto p_vec = chunk.release<PartitionVectorChunk>();
+        // for each local pid, it should receive num_chunks * nranks chunks.
+        EXPECT_EQ(p_vec.data.size(), num_chunks * n_ranks);
```
Comment (pentschev) [inline cpp/tests/streaming/test_shuffler.cpp]: This should be `ASSERT_EQ`. If the actor drops or duplicates chunks, the test continues and indexes `p_vec.data[r * num_chunks + c_idx]` below, which can go out of bounds instead of failing cleanly at the size mismatch.

File: cpp/tests/streaming/test_shuffler.cpp [hunk 5]
```diff
@@ -175,12 +78,111 @@ INSTANTIATE_TEST_SUITE_P(
...
+                    end - start, base + static_cast<int64_t>(start), stream, *br
+                )
+            );
+            cd.content_size(it->second.data->mem_type()) += it->second.data->size;
+        }
+        input_chunks.emplace_back(Message(
+            chunk_idx,
+            std::make_unique<PartitionMapChunk>(std::move(chunks)),
+            std::move(cd)
+        ));
```
Comment (pentschev) [inline cpp/tests/streaming/test_shuffler.cpp]: Could this use `to_message(chunk_idx, std::make_unique<PartitionMapChunk>(std::move(chunks)))` instead of constructing `Message` manually? The manual `ContentDescription cd{}` leaves the message non-spillable and omits the copy callback, so this no longer exercises the normal `PartitionMapChunk` streaming message path through `SpillableMessages`.

**Review by pentschev — CHANGES_REQUESTED** [review]

File: cpp/tests/streaming/test_shuffler.cpp (hunk 5 above)
Comment (nirandaperera) [reply cpp/tests/streaming/test_shuffler.cpp]: Oh yes. Good idea!

File: cpp/include/rapidsmpf/utils/misc.hpp (hunk 2 above)
Comment (nirandaperera) [reply cpp/include/rapidsmpf/utils/misc.hpp]: Done

File: cpp/tests/test_shuffler.cpp [hunk 6]
```diff
@@ -130,116 +130,135 @@ MemoryLimitsMap get_memory_limits_map(rapidsmpf::MemoryType priorities) {
...
+// are empty). Sub-region (local_pidx, split_idx) is piece k = local_pidx*P + split_idx
+// and is routed to destination partition split_idx; input region local_pidx is the union
+// of its P sub-regions. The pieces exactly tile [0,N), so the total shuffled data == N
+// regardless of rank/partition counts (conservation). A per-shuffle `base` offset is
+// added to every value so distinct shuffles carry distinct data.
+
+// Produces the non-empty sub-regions of one owned input region `local_pidx`, keyed by
+// destination partition. Since local_partitions() across ranks partition [0,P), every
+// input region is produced exactly once, so rows are not replicated.
+std::unordered_map<rapidsmpf::shuffler::PartID, rapidsmpf::PackedData>
```
Comment (madsbk) [inline cpp/tests/test_shuffler.cpp:173]: Use a normal docstring format explaining all `@param`

File: cpp/tests/test_shuffler.cpp [hunk 7]
```diff
@@ -130,116 +130,135 @@ MemoryLimitsMap get_memory_limits_map(rapidsmpf::MemoryType priorities) {
...
+                )
+            );
+        }
+    }
+    return chunks;
+}
+
+// Verifies that the `received` chunks for partition `j` match the non-empty sub-regions
+// expected for it.
+void validate_partition_data(
```
Comment (madsbk) [inline cpp/tests/test_shuffler.cpp:218]: Use a normal docstring format explaining all `@param`

File: python/rapidsmpf/rapidsmpf/testing.py [hunk 8]
```diff
@@ -6,13 +6,23 @@
 
 from typing import TYPE_CHECKING
 
+import numpy as np
 import pylibcudf
```
Comment (madsbk) [inline python/rapidsmpf/rapidsmpf/testing.py:10]: Remove?

Comment (pentschev) [reply python/rapidsmpf/rapidsmpf/testing.py:10]: Not ready for that yet. This PR adds new tests, removal of cudf dependencies need to be handled separately as a follow-up to https://github.com/rapidsai/cudf/pull/22814 in RapidsMPF.

File: python/rapidsmpf/rapidsmpf/testing.py [hunk 9]
```diff
@@ -63,3 +73,178 @@ def assert_eq(
...
+        CUDA stream used for the device allocation.
+    br
+        Buffer resource used for memory management.
+
+    Returns
+    -------
+    A ``PackedData`` containing the integer sequence.
+    """
+    data = np.arange(offset, offset + n_elements, dtype=_DTYPE).tobytes()
+    gpu_data = rmm.DeviceBuffer.to_device(data, stream=stream)
```
Comment (pentschev) [inline python/rapidsmpf/rapidsmpf/testing.py]: Please allocate this test payload through `br.device_mr` (or a small BR-backed helper) before wrapping it in `PackedData`. `to_device()` uses the current/default RMM resource and `from_device_buffer()` only adopts the allocation, so the replacement tests bypass `BufferResource` tracking that the old `partition_and_pack(..., br=br)` path covered.

File: python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py [hunk 10]
```diff
@@ -3,290 +3,95 @@
...
-            num_partitions,
-            partition_assignment=PartitionAssignment.CONTIGUOUS,
+    """
+    End-to-end correctness of the async streaming shuffler.
+
+    Each rank inserts the input regions it owns and, after shuffling, every local
+    partition is validated against the conserved data model.
+    """
+    br = context.br()
+    shuffler = ShufflerAsync(context, comm, 0, total_num_partitions)
```
Comment (pentschev) [inline python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py:55]: This replacement loses Python coverage of the streaming actor wrapper. The old cudf-based test drove the channel pipeline through `rapidsmpf.streaming.coll.shuffler.shuffler()`, while this now instantiates `ShufflerAsync` directly and never exercises the wrapper or its `PartitionAssignment.CONTIGUOUS` branch. Please add a cudf-free actor/channel test, or expose a generic `PartitionMapChunk` factory so the new `PackedData` model can drive `shuffler()`, ideally covering both assignment modes.

**Review by pentschev — CHANGES_REQUESTED** [review]

File: python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py (hunk 10 above)
Comment (nirandaperera) [reply python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py:55]: Good call!

File: python/rapidsmpf/rapidsmpf/testing.py (hunk 9 above)
Comment (nirandaperera) [reply python/rapidsmpf/rapidsmpf/testing.py]: Done

File: python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py (hunk 10 above)
Comment (nirandaperera) [reply python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py:55]: I ported the tests over.

**Review by madsbk — APPROVED** [review]
Thanks @nirandaperera

File: cpp/tests/test_shuffler.cpp [hunk 11]
```diff
@@ -130,116 +141,181 @@ MemoryLimitsMap get_memory_limits_map(rapidsmpf::MemoryType priorities) {
...
+/// Conservation-preserving data model shared by the shuffler round-trip tests.
+///
+/// The index range `[0, total_num_rows)` is split into `total_num_partitions^2`
+/// contiguous sub-regions via `chunk_indices` (front-loaded, so trailing sub-regions
+/// are empty when `N < P*P`). Sub-region `(local_pidx, split_idx)` is piece
+/// `k = local_pidx * P + split_idx` and is routed to destination partition
+/// `split_idx`. The pieces exactly tile `[0, N)`, so the total shuffled row
+/// count equals `N` regardless of rank or partition counts (conservation). A
+/// per-shuffle `base` offset is added to every value so distinct shuffles carry
+/// distinct data.
```
Comment (pentschev) [inline cpp/tests/test_shuffler.cpp:153]: Is this supposed to be part of `make_partition_data` docs? Currently looks like a strange placing not really connected to anything.

File: python/rapidsmpf/rapidsmpf/streaming/chunks/partition.pyx [hunk 12]
```diff
@@ -27,6 +95,62 @@ cdef class PartitionMapChunk:
...
+
+        Returns
+        -------
+        A new PartitionMapChunk owning the given packed data.
+
+        Raises
+        ------
+        ValueError
+            If any of the provided PackedData objects is empty.
+        """
```
Comment (pentschev) [inline python/rapidsmpf/rapidsmpf/streaming/chunks/partition.pyx:120]: The stub advertises `Mapping`/`Sequence` for these new helpers, but the Cython signatures require concrete `dict`/`list`. That means callers accepted by the type hints fail at runtime, for example a `MappingProxyType` or tuple. Since these helpers are now public in the `.pyi`, please either loosen the Cython parameters to accept the advertised protocols or narrow the stub to `dict[int, PackedData]` / `list[PackedData]`. The same applies to `from_packed_data_list`.

File: python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py [hunk 13]
```diff
@@ -155,8 +137,8 @@ async def generate_inputs(
 async def do_shuffle(
     context: Context,
     comm: Communicator,
-    ch_in: Channel[TableChunk],
-    ch_out: Channel[TableChunk],
+    ch_in: Channel[PartitionMapChunk],
+    ch_out: Channel[PartitionVectorChunk],
```
Comment (pentschev) [inline python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py]: I can't add the comment on the exact lines, but assumes this extends to line 146:

This replacement still bypasses the exported streaming actor wrapper. The removed cudf-based test drove `rapidsmpf.streaming.coll.shuffler.shuffler()` in the channel pipeline ([see that in 26.06 for example](https://github.com/rapidsai/rapidsmpf/blob/78b66f1a8ab6068cd38d73429e3935c3cb97c90f/python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py#L98-L108)), which exercised `streaming/coll/shuffler.pyx` including the `cpp_shuffler` call and `PartitionAssignment` mapping. This new actor instantiates `ShufflerAsync` directly, so a regression in the public `shuffler()` wrapper would no longer be caught. Now that `PartitionMapChunk.from_packed_data_map()` exists, can this test use the actual `shuffler()` actor between `ch_in` and `ch_shuffled`?

**Review by pentschev — CHANGES_REQUESTED** [review]

**Comment by pentschev** [comment]
I have opened #1092 and #1093 adding cuDF-free coverage for the remaining missing tests. Both require changes from this PR (specifically, `rapidsmpf.testing.generate_packed_data` and `rapidsmpf.testing.validate_packed_data`). _I think_ after this PR #1084 plus the aforementioned PRs are merged we would have covered everything missing and could then remove all remaining cuDF pieces without functionality coverage loss.

File: cpp/tests/test_shuffler.cpp (hunk 11 above)
Comment (nirandaperera) [reply cpp/tests/test_shuffler.cpp:153]: I was trying to explain test approach here, especially the `P^2` splits because I felt that a new reader might find it a little confusing.

Comment (nirandaperera) [reply cpp/tests/test_shuffler.cpp:153]: Since this relates to make_partition_data, validate_partition_data, test_shuffler methods I thought this is the best place to put it, more as a detail rather than a doc string.

File: python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py (hunk 13 above)
Comment (nirandaperera) [reply python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py]: Yes you are right. I was more focused on porting the previous tests as is, and missed this detail. I removed the `do_shuffle` actor alltogether, because we can strictly replace it from `shuffle` actor as you said.

File: python/rapidsmpf/rapidsmpf/streaming/chunks/partition.pyx (hunk 12 above)
Comment (nirandaperera) [reply python/rapidsmpf/rapidsmpf/streaming/chunks/partition.pyx:120]: done

File: python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py [hunk 14]
```diff
@@ -222,16 +168,27 @@ def test_shuffler_runtime_obeys_contiguous_assignment(
...
-    expected_local = list(
-        range(
-            rank * num_partitions // nranks,
-            (rank + 1) * num_partitions // nranks,
-        )
-    )
-    assert set(received_pids) == set(expected_local)
-    assert len(received_pids) == len(expected_local)
+    # Single rank, so every partition is local to this rank.
+    assert set(received_pids) == set(range(num_partitions))
```
Comment (pentschev) [inline python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py:172]:
```suggestion
    assert set(received_pids) == set(range(num_partitions))
    assert len(received_pids) == num_partitions
```

Comment (pentschev) [reply python/rapidsmpf/rapidsmpf/tests/streaming/test_shuffler.py:172]: Checks for duplicates, like previously.

**Review by pentschev — APPROVED** [review]
Thanks @nirandaperera , one final minor request, otherwise LGTM.

**Comment by nirandaperera** [comment]
Thanks @pentschev

File: cpp/tests/test_shuffler.cpp (hunk 11 above)
Comment (pentschev) [reply cpp/tests/test_shuffler.cpp:153]: It's a strange placement, but I don't have a better idea where to put it either. Let's leave it there for now.
